dashboard/app/controllers/api/v1/sections_students_controller.rb

```diff
@@ -24,6 +24,26 @@ def update
     end
   end
 
+  # Used only for picture and word sections
+  # POST /sections/<section_id>/students/bulk_add
+  def bulk_add
+    new_students = []
+    params[:students].each do |student|
```

Comment:
@aoby since this is in a loop, I'm not sure how to do a check for errors like I did in update. I also couldn't find any examples of `collection` `post`s to follow. Do you know of any?

Reply:
I'm guessing that we'd actually want the entire thing to fail if there are any errors, right? In that case, you'd probably want to wrap the whole thing in a transaction, in a block something like:

```
errors = []
ActiveRecord::Base.transaction do
  params[:students].each do |student|
    begin
      #existing inner loop logic goes here
    rescue Exception => e
      errors << e.message
    end
  end
  raise ActiveRecord::Rollback if errors.count > 0 # rolls back the transaction, but will not propagate
end
```

and then render some error information.

Reply:
I like @ewjordan 's suggested transaction flow, but there is no need to rescue anything inside the loop, and rescuing `Exception` is strongly discouraged in any case *. 

Although we prefer the bang ActiveRecord methods (e.g. `create!`) in many contexts to ensure we don't silently ignore errors, a controller that _expects_ and _handles_ the failure case is a perfect place to use the non-bang method. Just like we branched on the result of `update` in our [update action](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/api/v1/sections_students_controller.rb#L20-L24), we can here branch on the result of [create](https://apidock.com/rails/v4.2.7/ActiveRecord/Persistence/ClassMethods/create). 

Then you have a few choices on how to handle errors:
 1. As Eric suggests, attempt all, rollback and fail if you encounter any error. 
 2. Abort and rollback on the first error. 
 3. Attempt to create each student, collecting errors for the failed ones. This way if you are creating 100 students and only 1 fails, you still create the 99. This way you won't need a transaction, but you'll have to handle hybrid results.
 

--------
* `Exception` is the root of Ruby's [exception hierarchy](http://rubylearning.com/satishtalim/ruby_exceptions.html) so rescuing that can hide errors like `fatal` or `NoMemoryError`.  Generally, the least specific error you should rescue is `StandardError` but it's always better to rescue something more specific if you can.